ndsl/testing/comparison.py

```diff
@@ -1,68 +1,265 @@
+        # Summary and worst result
+        fullcount = len(self.references.flatten())
+        report.append(
+            f"Failed count: {bad_indices_count}/{fullcount} "
+            f"({round(100.0 * (bad_indices_count / fullcount), 2)}%),\n"
+            f"Worst failed index {worst_full_idx}\n"
+            f"    Computed:{computed_worst}\n"
+            f"    Reference: {reference_worst}\n"
+            f"    Absolute diff: {worst_abs_err:.3e}\n"
+            f"    Metric diff: {worst_metric_err:.3e}\n"
+            f"    Metric threshold: {self.eps}\n"
+            f"  Noise quantification:\n"
+            f"    Reference dtype: {type(reference_worst)}\n"
+            f"    Unique errors: {unique_errors}/{bad_indices_count}"
         )
-    )
+        report.extend(details)
+
+        return "\n".join(report)
+
+
+class MultiModalFloatMetric(BaseMetric):
+    """Combination of absolute, relative & ULP comparison for floats
+
+    This metric attempts to combine well known comparison on floats
+    to leverage a robust 32/64 bit float comparison on large accumulating
+    floating errors.
+
+    ULP is used to clear noise (ULP<=1.0 passes)
+    Absolute errors for large amplitute
+    """
+
+    _f32_absolute_eps = 1e-10
+    _f64_absolute_eps = 1e-13
+
+    def __init__(
+        self,
+        reference_values: np.ndarray,
+        computed_values: np.ndarray,
+        eps: float,
+        **kwargs,
+    ):
+        super().__init__(reference_values, computed_values)
+        self.absolute_distance = np.empty_like(self.references)
+        self.absolute_distance_metric = np.empty_like(self.references, dtype=np.bool_)
+        self.relative_distance = np.empty_like(self.references)
+        self.relative_distance_metric = np.empty_like(self.references, dtype=np.bool_)
+        self.ulp_distance = np.empty_like(self.references)
+        self.ulp_distance_metric = np.empty_like(self.references, dtype=np.bool_)
+
+        self.relative_fraction = 0.000001
+        if self.references.dtype is (np.float32, np.int32):
+            self.absolute_eps = max(eps, self._f32_absolute_eps)
+        else:
+            self.absolute_eps = max(eps, self._f64_absolute_eps)
+        self.ulp_threshold = 1.0
+
+        self.success = self._compute_all_metrics()
+        self.check = np.all(self.success)
+
+    def _compute_all_metrics(
+        self,
+    ) -> npt.NDArray[np.bool_]:
+        if self.references.dtype in (np.float64, np.int64, np.float32, np.int32):
+            max_values = np.maximum(
+                np.absolute(self.computed), np.absolute(self.references)
+            )
+            # Absolute distance
+            self.absolute_distance = np.absolute(self.computed - self.references)
+            self.absolute_distance_metric = self.absolute_distance < self.absolute_eps
+            # Relative distance (in pct)
+            self.relative_distance = np.divide(self.absolute_distance, max_values)
```

Comment:
we don't have the div by 0 check in here that is in the legacy. Not needed?

Reply:
Well we have no denominator normalization here so no risk of div by 0. The NaNs are checked independently.